### PR TITLE
Add opt-in structured JSONL logging to the NScript core pipeline

### PR DESCRIPTION
**Pull request by gautam-achieveai**
## Summary

Implements #10 — opt-in structured JSONL logging for the two-stage NScript compiler pipeline. When `--log <path>` is supplied to either stage, a single shared Serilog/CompactJsonFormatter file sink captures events from both `csc` (Stage 1) and `cs2jsc` (Stage 2). When omitted, behavior is unchanged — no logger bootstrap, no file I/O.

## Changes

- **`NScript.Utils/CompilerLog.cs`** — new static facility wrapping Serilog; idempotent `Initialize`/`Shutdown`; `ForComponent()` returns a silent no-op logger when disabled. Enriches events with `RunId`, `Stage`, `Pid`, `MachineName`.
- **Stage 1 (`CscCompiler.Main`)** — strips `--log`/`--run-id` from argv before `DesktopBuildClient.Run` (Roslyn would reject unknown switches). Supports `--log path`, `--log:path`, `--log=path`, and `-log`/`/log` aliases.
- **Stage 2 (`NScriptCompiler.Compile`, `ParseOptions`)** — accepts the same flags, resolves through `CompilerLog.Initialize`.
- **Env var fallbacks** — `NSCRIPT_LOG_PATH` and `NSCRIPT_LOG_RUNID` for MSBuild response-file scenarios.
- **Stage events** emitted from `Builder`, `ConverterContext`, `XwmlTemplatingPlugin`, `SerializationHelper` (all gated by `CompilerLog.IsEnabled` to avoid allocation when logging is disabled).
- **`RazorSkinCompiler.Logger`** — now routes through `CompilerLog.ForComponent("RazorSkinParser")`. The hard-coded `logs/razor-skin-compiler.log.jsonl` sink is removed — consumers must now opt in via `--log`.
- **ADR 0025** documents the design decision; **README** gains a Structured Logging section.
- **12 unit tests** cover the facility and the Stage 1 flag extraction.

Each JSONL event carries `@t`, `@l`, `Component`, `Stage`, `RunId`, `Pid`, `MachineName`.

### Cross-stage correlation example

```bash
csc.exe Program.cs --log build.jsonl --run-id my-build-001
NScript -outJs app.js -references Foo.dll -entryAssembly App.dll --log build.jsonl --runid my-build-001
```

Both stages append to `build.jsonl`; records can be joined on `RunId`.

## Breaking change

`RazorSkinParser`'s previously hard-coded `logs/razor-skin-compiler.log.jsonl` file is no longer produced. Consumers must now pass `--log` (or set `NSCRIPT_LOG_PATH`). Called out in the ADR and README.

## Test plan

- [x] New `CompilerLogTests` (8) — Initialize idempotency, silent-logger default, env-var fallback, missing-directory creation, shutdown flushes
- [x] New `CscFlagExtractionTests` (4) — space/colon/equals delimited flags, no-flag passthrough, null args
- [x] All 12 new tests pass
- [x] Existing `RazorSkinParser.Test` (164), `NScript.Converter.Test` (401), `CssParser.Test`, `NScript.JSParser.Test` all pass
- [x] Solution builds clean in both Release and Debug
- [x] Manual trace: running `cs2jsc --log build.jsonl --runid test` produces JSONL with expected fields

### CODE REVIEW

File: Sources/Compiler/NScript.Lib/ParseOptions.cs [hunk 1]
```diff
@@ -207,6 +227,32 @@ public static ParseOptions ParseArgs(string[] args)
                     case "-optimize":
                         options.optimize = true;
                         continue;
+                    case "-log":
```
Comment (gautam-achieveai) [inline Sources/Compiler/NScript.Lib/ParseOptions.cs:248]: [Gautam's review bot] F1: CLI Flag Parsing Inconsistency (HIGH)

Stage-2 parsing only accepts space-delimited `-log path` and `--log path`. But Stage-1 (`Csc.cs:ExtractNScriptFlags`) also accepts: `/log` (Windows-style slash), `--log:path` (colon-delimited), `--log=path` (equals-delimited).

A user who learns `--log:build.jsonl` works with csc will be confused when cs2jsc rejects it. This is the first NScript-specific flag, setting the precedent for all future flags.

Also: when `--log` is the last arg with no value, csc silently disables logging (logPath=null) while cs2jsc aborts with an error. These should behave the same.

**Suggested fix:** Either extract a shared `NScriptFlagParser` in `NScript.Utils`, or add colon/equals/slash support to this switch.

File: Sources/Compiler/NScript.Utils/CompilerLog.cs [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+                if (!isEnabled)
+                {
+                    return;
+                }
+
+                try
+                {
+                    Log.CloseAndFlush();
+                }
+                catch
```
Comment (gautam-achieveai) [inline Sources/Compiler/NScript.Utils/CompilerLog.cs]: [Gautam's review bot] F4: Bare catch in Shutdown (HIGH)

This bare `catch` silently discards all exceptions (including non-CLS) with zero diagnostic output. Compare with `Initialize()` at line 134 which properly uses `catch (Exception ex)` and writes to stderr.

If `CloseAndFlush()` fails, the user gets no signal about why their log file is truncated/empty.

**Fix:** `catch (Exception ex) { Console.Error.WriteLine($"CompilerLog: flush failed during shutdown ({ex.Message})."); }`

File: Sources/Compiler/RazorSkinParser/RazorSkinCompiler.cs [hunk 3]
```diff
@@ -1,27 +1,20 @@
...
-
         /// <summary>
-        /// Static logger accessible to other classes in the pipeline.
+        /// Static logger accessible to other classes in the pipeline. Routes through
+        /// the shared <see cref="CompilerLog"/> facility — returns a silent no-op
+        /// logger when <c>--log</c> has not been supplied to the host (csc/cs2jsc),
+        /// so no file I/O occurs by default.
         /// </summary>
-        public static ILogger Logger => Log;
+        public static ILogger Logger => CompilerLog.ForComponent("RazorSkinParser");
```
Comment (gautam-achieveai) [inline Sources/Compiler/RazorSkinParser/RazorSkinCompiler.cs:17]: [Gautam's review bot] F10: Logger Property Creates Allocation Per Access (HIGH)

`CompilerLog.ForComponent("RazorSkinParser")` calls `rootLogger.ForContext(...)` which allocates a new enriched logger wrapper every time this property is accessed. 11 classes chain through this with ~60+ call sites per template compilation.

This is the pattern being established for the module. Fixing later means touching all 11 files.

**Fix:** Cache the result: `private static ILogger _cachedLogger; public static ILogger Logger => _cachedLogger ??= CompilerLog.ForComponent("RazorSkinParser");`

File: Test/Compiler/NScript.Utils.Test/CompilerLogTests.cs [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        [TestMethod]
+        public void ExtractNScriptFlags_EmptyArgs_ReturnsEmpty()
+        {
+            var filtered = NScript.Csc.Lib.CscCompiler.ExtractNScriptFlags(null, out var logPath, out var runId);
+            Assert.IsNull(logPath);
+            Assert.IsNull(runId);
+            Assert.AreEqual(0, filtered.Length);
+        }
+    }
+}
```
Comment (gautam-achieveai) [inline Test/Compiler/NScript.Utils.Test/CompilerLogTests.cs:363]: [Gautam's review bot] F6: Missing ParseOptions Test Coverage (HIGH)

The `CscFlagExtractionTests` (4 tests) cover Stage-1 flag parsing well. But Stage-2 (`ParseOptions.ParseArgs`) has zero test coverage for the new `--log`/`--runid` branches — and its parsing logic is completely different.

This gap is how the flag inconsistency (F1) went undetected.

**Missing tests:** (1) ParseOptions with `--log path` + `--run-id id` (happy path), (2) `--log` as last arg (error path), (3) ExtractNScriptFlags flag at end of args with no value, (4) Duplicate `--log` flags (last-write-wins).

**Review by gautam-achieveai — COMMENTED** [review]
## [Gautam's review bot] PR #23 — Deep Review

**Verdict: REQUEST CHANGES** (4 blocking findings, 5 non-blocking, 1 pre-existing)

### Summary
Well-designed PR that introduces cross-cutting observability with minimal architectural disruption. The `CompilerLog` facility, opt-in design, and per-component naming are all sound choices. ADR 0025 is thorough. Build passes (0 errors), all 12 new tests pass.

However, four issues should be resolved before merge:

### Blocking Issues

| # | Severity | Finding | File(s) |
|---|----------|---------|---------|
| F1 | :red_circle: HIGH | **CLI flag parsing inconsistency** — csc accepts `/log`, `--log:path`, `--log=path`; cs2jsc only accepts space-delimited. Also: csc silently ignores `--log` with no value; cs2jsc aborts. | `Csc.cs` vs `ParseOptions.cs` |
| F4 | :red_circle: HIGH | **Shutdown() bare `catch`** swallows all exceptions silently, while Initialize() properly logs to stderr. Asymmetric error handling in a diagnostics facility. | `CompilerLog.cs:180` |
| F6 | :red_circle: HIGH | **Zero test coverage** for ParseOptions `--log`/`--runid` parsing (Stage-2 path). The csc side has 4 tests — this gap is how F1 went undetected. | `ParseOptions.cs` |
| F10 | :red_circle: HIGH | **`RazorSkinCompiler.Logger` property** creates a new `ForContext` wrapper on every access. 11 consumer classes x ~60 call sites per template = hundreds of short-lived allocations. This is the pattern being established — fixing later means touching all 11 files. | `RazorSkinCompiler.cs:17` |

### Non-Blocking Issues

| # | Severity | Finding | File(s) |
|---|----------|---------|---------|
| F2 | :yellow_circle: MEDIUM | `ForComponent()` TOCTOU race — `rootLogger` read without local snapshot | `CompilerLog.cs:151` |
| F3 | :yellow_circle: MEDIUM | `Log.Logger = rootLogger;` sets process-global Serilog static unnecessarily | `CompilerLog.cs:127` |
| F5 | :green_circle: LOW | Redundant `string.Split()` — same split computed twice in XwmlTemplatingPlugin | `XwmlTemplatingPlugin.cs:108` |
| F7 | :green_circle: LOW | Redundant direct Serilog `PackageReference` in RazorSkinParser.csproj (now transitive via NScript.Utils) | `RazorSkinParser.csproj` |
| F9 | :green_circle: LOW | Inconsistent `IsEnabled` guards — 6 sites guard, 2 don't. `ForComponent()` already returns silent logger; pick one convention. | Multiple files |

### Pre-Existing (FYI, not blocking)
- **TemplateIRBuilder.cs:168-181**: Unguarded debug loop runs for every template with expensive string building (StringBuilder + reflection) even when logging is disabled. Not introduced by this PR but made more visible by the logging migration. Recommend a follow-up issue.

### Positive Observations
- Roslyn integration boundary (ADR 0005) well preserved — flag stripping is clean
- Logging instrumentation respects component boundaries
- Static singleton appropriate for the two-process architecture
- ADR 0025 is well-written
- Initialize() catch block is exemplary infrastructure error handling
- Try/finally lifecycle pattern in both entry points is correct

### Review Agents Used
architecture-review, exception-handling-review, performance-review, test-coverage-review, code-simplifier, duplicate-code-detector, review-grader

---
*Linked to User Story #10*

File: Sources/Compiler/NScript.Utils/CompilerLog.cs [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Returns a logger enriched with the given <c>Component</c> name. When the
+        /// shared logger has not been initialized, returns a silent no-op logger.
+        /// </summary>
+        public static ILogger ForComponent(string component)
+        {
+            if (!isEnabled || rootLogger == null)
```
Comment (gautam-achieveai) [inline Sources/Compiler/NScript.Utils/CompilerLog.cs]: [Gautam's review bot] F2: ForComponent() TOCTOU Race (MEDIUM)

`rootLogger` is read at line 151 (null-check) then dereferenced at line 156, but a concurrent `Shutdown()` could set `rootLogger = null` between reads — causing `NullReferenceException`.

**One-line fix:** Snapshot to local: `var logger = rootLogger; if (!isEnabled || logger == null) { return silentLogger; } return logger.ForContext(...);`

File: Sources/Compiler/NScript.Utils/CompilerLog.cs [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                        .Enrich.WithProperty("MachineName", Environment.MachineName)
+                        .WriteTo.File(
+                            formatter: new CompactJsonFormatter(),
+                            path: resolvedPath,
+                            shared: true,
+                            rollOnFileSizeLimit: false,
+                            flushToDiskInterval: TimeSpan.FromSeconds(1));
+
+                    rootLogger = configuration.CreateLogger();
+                    Log.Logger = rootLogger;
```
Comment (gautam-achieveai) [inline Sources/Compiler/NScript.Utils/CompilerLog.cs:135]: [Gautam's review bot] F3: Unnecessary Global Side-Effect (MEDIUM)

`Log.Logger = rootLogger;` sets the Serilog process-global static logger. All consumer code uses `CompilerLog.ForComponent()` — nobody calls `Log.Information()` directly. This creates invisible coupling (any code using Serilog's `Log.*` API silently routes to the JSONL sink) and blocks test parallelism.

Consider removing this line and changing `Shutdown()` to dispose `rootLogger` directly.

File: Sources/Compiler/XwmlParser/XwmlTemplatingPlugin.cs [hunk 7]
```diff
@@ -98,6 +98,15 @@ public void Initialize(NScript.CLR.ClrContext clrContext, RuntimeScopeManager ru
...
 
             this.parserContext.ConverterContext = runtimeScopeManager.Context;
+
+            if (CompilerLog.IsEnabled)
+            {
+                CompilerLog.ForComponent("XwmlParser").Information(
+                    "XwmlParser.Initialized KnownCssClassCount={Count}",
+                    string.IsNullOrEmpty(this.knownCssClasses)
+                        ? 0
+                        : this.knownCssClasses.Split(new[] { ',', ' ' }, StringSplitOptions.RemoveEmptyEntries).Length);
```
Comment (gautam-achieveai) [inline Sources/Compiler/XwmlParser/XwmlTemplatingPlugin.cs]: [Gautam's review bot] F5: Redundant string.Split() (LOW)

The same `knownCssClasses.Split(...)` is called at line 98 for the `ParserContext` constructor and again here just for `.Length`. The second call allocates a throwaway string array.

**Fix:** Extract to local variable before both usages.

**Comment by gautam-achieveai** [comment]
## Addressed Review Feedback (53475e36)

All 4 blocking (HIGH) and 2 non-blocking findings have been resolved in the latest push.

### Blocking Issues - Fixed

| # | Finding | Fix |
|---|---------|-----|
| **F1** | CLI flag parsing inconsistency | Stage-2 `ParseOptions` now supports colon (`--log:path`), equals (`--log=path`), and slash (`/log`) forms, matching Stage-1. Missing-value behavior changed from abort to graceful disable (matching Stage-1). |
| **F4** | Bare `catch` in `Shutdown()` | Now `catch (Exception ex)` with stderr diagnostic, matching `Initialize()` pattern. |
| **F6** | Zero test coverage for Stage-2 flags | Added 7 new `ParseOptionsLogFlagTests`: space/colon/equals/slash forms, missing-value edge cases for both `--log` and `--run-id`, and no-flag baseline. |
| **F10** | Logger allocation per access | Added `ConcurrentDictionary` component cache in `CompilerLog.ForComponent()` with lifecycle-aware invalidation on `Shutdown()`. All callers benefit without code changes. |

### Non-Blocking - Fixed

| # | Finding | Fix |
|---|---------|-----|
| **F2** | `ForComponent()` TOCTOU race | `rootLogger` now snapshotted to local before null-check and dereference. |
| **F5** | Redundant `string.Split()` | Extracted to local variable used by both the `ParserContext` ctor and the log statement. |

### Non-Blocking - Deferred

| # | Finding | Reason |
|---|---------|--------|
| **F3** | Unnecessary `Log.Logger` global | Removing requires repo-wide audit. Low risk since all new code uses `CompilerLog.ForComponent()`. |
| **F7** | Redundant Serilog PackageReference | Pre-existing transitive dependency; low priority. |
| **F9** | Inconsistent `IsEnabled` guards | Redundant but harmless; can be standardized in follow-up. |

### Test Results
- All 19 `NScript.Utils.Test` tests pass (8 CompilerLog + 4 CscFlag + 7 new ParseOptions)
- All 401 `NScript.Converter.Test` tests pass
- All 164 `RazorSkinParser.Test` tests pass
- Solution builds clean (0 errors)

File: Sources/Compiler/NScript.Utils/CompilerLog.cs [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+            }
+
+            var key = component ?? "unknown";
+            if (componentCache.TryGetValue(key, out var cached))
+            {
+                return cached;
+            }
+
+            var enriched = logger.ForContext("Component", key);
+            componentCache.TryAdd(key, enriched);
```
Comment (gautam-achieveai) [inline Sources/Compiler/NScript.Utils/CompilerLog.cs:174]: [Gautam's review bot] **F11: Non-blocking suggestion — use `GetOrAdd` instead of `TryGetValue` + `TryAdd`** (MEDIUM)

The current two-step pattern has a benign race: when two threads first call `ForComponent("Converter")` concurrently, both miss `TryGetValue`, both allocate a `ForContext` wrapper, and only the first `TryAdd` wins. The losing thread returns its own non-cached instance — a different object than what's in the cache.

`ConcurrentDictionary.GetOrAdd` is the standard single-call API for this pattern. Both threads get back the same winning instance, and the code shrinks from 7 lines to 1:

```csharp
public static ILogger ForComponent(string component)
{
    var logger = rootLogger;
    if (!isEnabled || logger == null)
    {
        return silentLogger;
    }

    var key = component ?? "unknown";
    return componentCache.GetOrAdd(key, k => logger.ForContext("Component", k));
}
```

**Impact:** Low in practice — only ~6 component names, race only on first access, `ForContext` is lightweight. The win is simplicity and idiomatic correctness.

*Not blocking — take it or leave it.*

**Review by gautam-achieveai — COMMENTED** [review]
[Gautam's review bot] **Re-review of iteration 2 — APPROVE** ✅

All 4 blocking findings from the initial review have been thoroughly addressed:

| # | Finding | Status |
|---|---------|--------|
| **F1** | CLI flag parsing inconsistency | ✅ `TryExtractInlineValue` added for colon/equals; `/log` added to switch; missing-value behavior aligned with Stage 1 |
| **F4** | Bare `catch` in Shutdown | ✅ `catch (Exception ex)` with stderr diagnostic |
| **F6** | Missing ParseOptions tests | ✅ 7 new tests covering space/colon/equals/slash/missing-value/baseline |
| **F10** | Logger allocation per access | ✅ `ConcurrentDictionary` cache centralized in `ForComponent()` — better than the originally suggested per-consumer caching |

Non-blocking fixes also verified: F2 (TOCTOU snapshot ✅), F5 (redundant Split extracted ✅). Deferred items (F3, F7, F9) are acceptable.

**Tests:** 19/19 pass (8 CompilerLog + 4 CscFlag + 7 new ParseOptions)

**One new non-blocking suggestion** posted inline (F11: `GetOrAdd` simplification).

**Review agents used:** performance-review, exception-handling-review. Both confirmed the thread-safety patterns are correct, the TOCTOU fix is complete, and the Shutdown lifecycle is sound.

Good work addressing the feedback. The centralized `ConcurrentDictionary` cache in `ForComponent()` is a cleaner solution than per-consumer caching — all 11 `RazorSkinCompiler` consumers benefit automatically without code changes.
